tokio-rustls/src/server/async_acceptor.rs

```diff
@@ -0,0 +1,98 @@
+/// let mut acceptor = tokio_rustls::AsyncAcceptor::new(stream);
+/// match acceptor.accept().await {
+///     Ok(start) => {
+///         let config = choose_server_config(start.client_hello()).await.unwrap();
+///         let stream = start.into_stream(config).await.unwrap();
+///         // Proceed with handling the ServerConnection...
+///     }
+///     Err(err) => {
+///         if let Some(mut stream) = acceptor.take_io() {
+///             stream
+///                 .write_all(
+///                     format!("HTTP/1.1 400 Invalid Input\r\n\r\n\r\n{:?}\n", err)
+///                         .as_bytes()
+///                 )
+///                 .await
+///                 .unwrap();
+///         }
+///     }
+/// }
+/// ```
+pub struct AsyncAcceptor<IO> {
+    acceptor: rustls::server::Acceptor,
+    io: Option<IO>,
+}
+
+impl<IO> AsyncAcceptor<IO>
+where
+    IO: AsyncRead + Unpin,
+{
+    /// Return an empty Acceptor, ready to receive bytes from a new client connection (io).
+    #[inline]
+    pub fn new(io: IO) -> Self {
+        Self {
+            acceptor: rustls::server::Acceptor::default(),
+            io: Some(io),
+        }
+    }
+
+    /// Wait for the `ClientHello` message from client. Do not call this function more than once.
+    ///
+    /// Returns `Ok(accepted)` if the connection has been accepted.
+    ///
+    /// Returns `Err(err)` if an error occurred. Use [`take_io`] to retrieve the client connection.
+    pub async fn accept(&mut self) -> Result<StartHandshake<IO>, rustls::Error> {
+        poll_fn(|cx: &mut Context<'_>| loop {
+            let io = match self.io.as_mut() {
+                Some(io) => io,
+                None => {
+                    return Poll::Ready(Err(rustls::Error::General(
+                        "acceptor cannot be polled after acceptance".into(),
+                    )))
+                }
+            };
+
+            let mut reader = common::SyncReadAdapter { io, cx };
+            match self.acceptor.read_tls(&mut reader) {
+                Ok(0) => return Poll::Ready(Err(rustls::Error::HandshakeNotComplete)),
+                Ok(_) => {}
+                Err(e) if e.kind() == io::ErrorKind::WouldBlock => return Poll::Pending,
+                Err(e) => return Poll::Ready(Err(rustls::Error::General(e.to_string()))),
+            }
+
+            match self.acceptor.accept() {
+                Ok(Some(accepted)) => {
+                    let io = self.io.take().unwrap();
+                    return Poll::Ready(Ok(StartHandshake { accepted, io }));
+                }
+                Ok(None) => continue,
+                Err(err) => {
+                    return Poll::Ready(Err(err));
+                }
+            }
+        })
+        .await
+    }
+    /// Takes back the client connection. Will return `None` if called more than once or if the
+    /// connection has been accepted.
+    pub fn take_io(&mut self) -> Option<IO> {
```

Comment:
I don't see why we can't add a `take_io()` method to the `LazyConfigAcceptor`? I don't think adding a new type is the right direction here.

Reply:
I didn't know how to deal with fact that the `new` function works like an async function. I guess I can add a `take_io` method even so?  I didn't try that (I'm a brand new rust programmer).

Reply:
I just gave it a go but the problem is that the `acceptor.await` takes ownership of `acceptor` so I can't call the `take_io` method 🤔

Reply:
You can try

```rust
pin_mut!(acceptor);
let ret futures::future::poll_fn(|cx| acceptor.as_mut().poll(cx)).await;
```

Reply:
That worked! I didn't need to call `poll_fn`

```rust
           let mut acceptor = LazyConfigAcceptor::new(rustls::server::Acceptor::default(), stream);
            pin_mut!(acceptor);

            match acceptor.as_mut().await {
            ...
                Err(err) => {
                    if let Some(mut stream) = acceptor.take_io() {
                        stream
                            .write_all(
                                /* fdfdsfds fdffd */
                                format!("HTTP/1.1 500 Internal Server Error\r\n\r\n\r\n{:?}\n", err) /*        */
                                    .as_bytes(),
                            )
                            .await
                            .unwrap();
                    }
                }

            }
```


I'll close this and make another PR.